tests/unit/parallel/algorithms/stable_partition_tests.hpp

```diff
@@ -0,0 +1,285 @@
+    std::copy(boost::begin(c), boost::end(c), boost::begin(d));
+
+    int partition_at = std::rand();
+
+    auto f =
+        hpx::parallel::stable_partition(p,
+            iterator(boost::begin(c)), iterator(boost::end(c)),
+            less_than(partition_at));
+
+    auto result = f.get();
+    auto partition_pt = std::find_if(boost::begin(c), boost::end(c),
+        great_equal_than(partition_at));
+    HPX_TEST(result.base() == partition_pt);
+
+    // verify values
+    std::stable_partition(boost::begin(d), boost::end(d), less_than(partition_at));
+
+    std::size_t count = 0;
+    HPX_TEST(std::equal(boost::begin(c), boost::end(c), boost::begin(d),
+        [&count](std::size_t v1, std::size_t v2) -> bool {
+            HPX_TEST_EQ(v1, v2);
+            ++count;
+            return v1 == v2;
+        }));
+    HPX_TEST_EQ(count, d.size());
+}
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename ExPolicy, typename IteratorTag>
+void test_stable_partition_exception(ExPolicy policy, IteratorTag)
+{
+    static_assert(
+        hpx::parallel::is_execution_policy<ExPolicy>::value,
+        "hpx::parallel::is_execution_policy<ExPolicy>::value");
+
+    typedef std::vector<int>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<int> c(10007);
+    std::vector<int> d(c.size());
+    std::iota(boost::begin(c), boost::end(c), std::rand());
+
+    bool caught_exception = false;
+    try {
+        hpx::parallel::stable_partition(policy,
+            iterator(boost::begin(c)), iterator(boost::end(c)),
+            throw_always());
+
+        HPX_TEST(false);
+    }
+    catch(hpx::exception_list const& e) {
+        caught_exception = true;
+        test::test_num_exceptions<ExPolicy, IteratorTag>::call(policy, e);
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_exception);
+}
+
+template <typename ExPolicy, typename IteratorTag>
+void test_stable_partition_exception_async(ExPolicy p, IteratorTag)
+{
+    typedef std::vector<int>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<int> c(10007);
+    std::vector<int> d(c.size());
+    std::iota(boost::begin(c), boost::end(c), std::rand());
+
+    bool caught_exception = false;
+    bool returned_from_algorithm = false;
+    try {
+        auto f =
+            hpx::parallel::stable_partition(p,
+                iterator(boost::begin(c)), iterator(boost::end(c)),
+                throw_always());
+        returned_from_algorithm = true;
+        f.get();
+
+        HPX_TEST(false);
+    }
+    catch(hpx::exception_list const& e) {
+        caught_exception = true;
+        test::test_num_exceptions<ExPolicy, IteratorTag>::call(p, e);
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_exception);
+    HPX_TEST(returned_from_algorithm);
```

Comment:
My point was that the algorithm is invoked and throws, but if the thread invoking the algorithm suspends immediately after calling it, the algorithm might throw before the `returned_from_algorithm` flag is set. This would make the test fail (once every few million runs) even though the actaul test itself was still correct